File: src/main/java/me/chriss99/CameraMatrix.java
package me.chriss99;

import org.joml.Matrix4f;
import org.joml.Vector3f;

public class CameraMatrix {
    float aspectRatio = 1;
    float FOV = 90;
    float zNear = 0;
    float zFar = 20000;

    Vector3f position = new Vector3f();
    float yaw = 0;
    float pitch = 0;
    float roll = 0;

    public Matrix4f generateMatrix() {
        Matrix4f viewMatrix = new Matrix4f();
        viewMatrix.m30(-position.x);
        viewMatrix.m31(-position.y);
        viewMatrix.m32(-position.z);

        Matrix4f yawMatrix = new Matrix4f();
        yawMatrix.m00((float) Math.cos(yaw));
        yawMatrix.m02((float) -Math.sin(yaw));
        yawMatrix.m20((float) Math.sin(yaw));
        yawMatrix.m22((float) Math.cos(yaw));

        Matrix4f pitchMatrix = new Matrix4f();
        pitchMatrix.m11((float) Math.cos(pitch));
        pitchMatrix.m12((float) Math.sin(pitch));
        pitchMatrix.m21((float) -Math.sin(pitch));
        pitchMatrix.m22((float) Math.cos(pitch));

        Matrix4f rollMatrix = new Matrix4f();
        rollMatrix.m00((float) Math.cos(roll));
        rollMatrix.m01((float) Math.sin(roll));
        rollMatrix.m10((float) -Math.sin(roll));
        rollMatrix.m11((float) Math.cos(roll));

        float f = (float) (1 / Math.tan(Math.toRadians(FOV)/2));
        float DNF = 1/(zNear-zFar);

        Matrix4f projectionMatrix = new Matrix4f();
        projectionMatrix.m00(aspectRatio * f);
        projectionMatrix.m11(f);
        projectionMatrix.m22(-2 * DNF);
        projectionMatrix.m32(1 + (2*zFar) * DNF);
        projectionMatrix.m23(1);
        projectionMatrix.m33(0);

        return projectionMatrix.mul(rollMatrix.mul(pitchMatrix.mul(yawMatrix.mul(viewMatrix))));
    }
}

File: src/main/kotlin/CameraMatrix.kt
package me.chriss99

import glm_.mat4x4.Mat4
import glm_.vec3.Vec3
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.tan

data class CameraMatrix(
    var aspectRatio: Float = 1f,
    var FOV: Float = 90f,
    var zNear: Float = 0f,
    var zFar: Float = 20000f,

    var position: Vec3 = Vec3(),
    var yaw: Float = 0f,
    var pitch: Float = 0f,
    var roll: Float = 0f
) {
    val viewMatrix: Mat4
        get() = Mat4().let {
            it.d0 = -position.x
            it.d1 = -position.y
            it.d2 = -position.z
            it
        }
    val yawMatrix: Mat4
        get() = Mat4().let {
            it.a0 = cos(yaw)
            it.a2 = -sin(yaw)
            it.c0 = sin(yaw)
            it.c2 = cos(yaw)
            it
        }
    val pitchMatrix: Mat4
        get() = Mat4().let {
            it.b1 = cos(pitch)
            it.b2 = sin(pitch)
            it.c1 = -sin(pitch)
            it.c2 = cos(pitch)
            it
        }
    val rollMatrix: Mat4
        get() = Mat4().let {
            it.a0 = cos(roll)
            it.a1 = sin(roll)
            it.b0 = -sin(roll)
            it.b1 = cos(roll)
            it
        }
    val projectionMatrix: Mat4
        get() = Mat4().let {
            val f = (1 / tan(Math.toRadians(FOV.toDouble()) / 2)).toFloat()
            val DNF = 1 / (zNear - zFar)

            it.a0 = aspectRatio * f
            it.b1 = f
            it.c2 = -2 * DNF
            it.d2 = 1 + (2 * zFar) * DNF
            it.c3 = 1f
            it.d3 = 0f
            it
        }

    fun generateMatrix(): Mat4 = projectionMatrix * rollMatrix * pitchMatrix * yawMatrix * viewMatrix
}